vizro-core/src/vizro/managers/_model_manager.py

```diff
@@ -80,39 +93,69 @@ def _get_models(
+        # dashboard tree instead of yielding.
+        # MS: This misses items created in pre-build (like selectors) and "behind the scene" items like ActionsChain
+        # (and possibly more)
+        # TODO: decide if alternative scheme with init in
+        # VizroBaseModel is satisfactory, luckily action chains are gone
         from vizro.models import VizroBaseModel
 
         if isinstance(model, VizroBaseModel):
-            yield model
+            if parent is None:
+                # AM rough notes:
+                # Populate root node of dashboard. Maybe dashboard should be actual tree root instead
+                # (so Tree("dashboard")) but note there's no way to attach data to it then.
+                self.__dashboard_tree.add(model, kind="dashboard")
+            else:
+                self.__dashboard_tree[parent].add(model, kind=field_name)
+
             for model_field in model.__class__.model_fields:
-                yield from self.__get_model_children(getattr(model, model_field))
+                self.__populate_tree(getattr(model, model_field), model, model_field)
         elif isinstance(model, Mapping):
             # We don't look through keys because Vizro models aren't hashable.
-            for child in model.values():
-                yield from self.__get_model_children(child)
+            for child_model in model.values():
+                self.__populate_tree(child_model, parent, field_name)
         elif isinstance(model, Collection) and not isinstance(model, str):
-            for child in model:
-                yield from self.__get_model_children(child)
+            for child_model in model:
+                self.__populate_tree(child_model, parent, field_name)
+        # TODO[MS]: does there not need to be a case for not being a model? I guess that should
+        # just be caught by mypy
 
-    def _get_model_page(self, model: Model) -> Page:  # type: ignore[return]
+    def _get_model_page(self, model: Model) -> Page:
         """Gets the page containing `model`."""
         from vizro.models import Page
 
         if isinstance(model, Page):
             return model
 
-        for page in cast(Iterable[Page], self._get_models(Page)):
-            if model in self.__get_model_children(page):  # type: ignore[operator]
-                return page
+        # Find the model's node in the tree and walk up to find the Page
+        current_node = self.__dashboard_tree.find(data_id=model.id)
+        if not current_node:
+            raise ValueError(f"Model with id='{model.id}' not found in dashboard tree")
+
+        while current_node.parent:
+            current_node = current_node.parent
+            if isinstance(current_node.data, Page):
+                return current_node.data
+
+        raise ValueError(f"Model with id='{model.id}' is not contained within any Page")
 
     @staticmethod
```

Comment:
Let's move this directly into the `VizroBaseModel` validator unless there's a good reason not to do so?

Reply:
Yes exactly.